# Limpieza de Datos

En la presente sección se procede a limpiar los datos brutos de la lectora óptica. Para poder visualizar mejor el trabajo a
continuación se presentan el formato *limpio* de los datos y se compara con una muestra de los datos brutos.

## Aspecto de los datos

Idealmente para un proceso adecuado, los datos deberían tener el siguiente aspecto

![](https://github.com/kamecon/TFM_Kschool/raw/master/data1.jpg)

En este caso, se podrían cargar los datos en un data frame o incluso una hoja de cálculo sin problemas. No obstante los datos
brutos suelen tener este aspecto

![](https://github.com/kamecon/TFM_Kschool/raw/master/data2.jpg)

La anterior es una muestra sesgada de los datos para precisamente mostrar los elementos a limpiar, pero no es una muestra
exhaustiva de los errores que nos podemos encontrar. Dichos errores son una combinación de errores humanos y de lectura de la lectora óptica. A continuación se enumeran algunos:

* Los usuarios dejan casillas en blanco
* Colocan mal alguno de los códigos
* No rellenan bien la casilla
* La lectora desplaza algunos campos
* La lectora interpreta mal algún código y lanza un símbolo en lugar del número
* etc 

Dado que parte de estos errores se corregían de forma manual, afortunadamente tenemos una amplia compresión de su orígen y
solución


## Pre-procesado

Antes de comenzar con el proceso de limpieza de datos que posteriormente sea parte del flujo de trabajo automático, es necesario manipular este archivo debido a características particulares que en principio no estarán presentes en futuras salidas de la lectora

Debido a que los impresos de las encuestas solo poseen opciones de grupos de la _A_ a la _C_, hay dos grupos que al no poseer estas siglas han dejado esa casilla en blanco y no es posible determinar su pertenencia.

En el proceso manual esto se resolvía al momento de pasar las encuestas por la lectora, dado que la muestra que se posee no ha sido pasada en el mismo orden que en el procesado original, no es posible determinar el grupo correspondiente a estos registros.

Se opta por elimnarlos de la muestra con esto se perderían 1pico registros.

El siguiente problema guarda relación con los postgrados. Al ser grupos únicos tienen en blanco la casilla correspondiente al grupo. No obstante son solo dos grupos y se encuentran identificados por el código de división y curso, así que en este caso solo habría que rellenar el grupo en dichos registros.

En las próximas encuestas esto dos problemas ya estarán resueltos, por lo que se podría proceder directamente a la limpieza

## Proceso

Se importan las librerías necesarias (de momento)

In [1]:
import pandas as pd

In [2]:
import numpy as np

Se cargan los datos

In [10]:
encu_1 = pd.read_csv('Data/encuesta.txt', sep=';', skiprows=1, header=None)
encu_1.head()

0  1  2    3    4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  \
0  2  1  A  0 2  0 6  05  06  04  08  07  03  05  07  07  09                   
1  2  1  A  00   006  07  07  06  06  06  08  07      08  08                   
2  2  1  A  002  006  10  10  10  10  10  10  10  10  10  10                   
3  2  1  A  002       08  07  07  06  07  08  07      07  08                   
4  2  1  A  002  006  09  08  08  08  09  09  08  09  08  08                   

   19  
0 NaN  
1 NaN  
2 NaN  
3 NaN  
4 NaN

In [14]:
encu_1.tail()

0  1  2    3    4   5   6   7   8   9   10  11  12  13  14  15  16  17  \
3724           250  223  08  08  07  08  08  08  09  07  08  08               
3725           250  223  05  05  04  05  04  05  05  05  05  05               
3726           250  223  04  04  04  05  04  07  05  03  03  04               
3727  1        250  223  05  04  04  05  06  06  07  09  08  07               
3728           250  223  07  07  06  09  04  09  08  09  07                   

      18  19  
3724     NaN  
3725     NaN  
3726     NaN  
3727     NaN  
3728     NaN

### Pre - procesado

Este es el trozo a elminar, se muestra tanto el registro anterior (1º fila) como el posterior (última fila) para mostrar que la columna 2 correspondiente al grupo se encuentra en blanco en todos los casos.

Nótese asimismo como la columna 3 correspondiente a la asignatura cambia en el 1º y último registro

In [11]:
encu_1.iloc[3394:3509, :]

0  1  2    3    4   5   6   7   8   9   10  11  12  13  14  15  16  17  \
3394  4  4  C  422  431  09  10  09  08  09  10  10  10  10  10               
3395  4  4     420  430  09  09  09  10  09  08  09  10  09  09               
3396  4  4     420  430  09  09  09  09  07  06  09  09  09  09               
3397  4        420  430  10  10  10  10  10  09  10  09  09  10               
3398  4  4     420  430  07  07  08  08  03  06  09  09  09  07               
3399  4  4     420  430  07  08  07  07  06  06  08  07  08  08               
3400  4  4     420  430  07  08  07  08  09  07  08  07  07  07               
3401  4  4               08  07  10  10  09  10  10  10  08  09               
3402  4        419  446  08  04  06  08  07  09  08      05  05               
3403  4  4     419  446  08  09  07  09  05  08  09  05  09  08               
3404  4  4     419  446  10  07  08  06  08  10  10  07  10  09               
3405  4  4     419  446  06  06  06  07  06  08  06  07  06  07               
3406     4               10  10  10  10  09  10  10  08  10  10               
3407  4  4     419  429  08  09  08  08  08  09  10      09  09               
3408  4  4     419  429  09  10  07  09  05  09  09  06  09  09               
3409  4  4     419  446  09  09  09  08  09  09  09  09  10  08               
3410  4  4     419  429  07  10  08  10  09  09  10  05  10  09               
3411  4  4     418  428  06  05  05  09  06  10  10  10  07  09               
3412  4  4     418  428  06  08  07  08  08  09  09  06  08  08               
3413  4        418  428  08  09  07  06  09  07  09  08  07  09               
3414           418  428  07  05  07  09  08  10  09  08  08  08               
3415           418  428  07  07  08  09  07  06  06  07  05  06               
3416  4  4     418  428  06  06  06  07  06  07  08  05  07  07               
3417  4  4     418  428  06  06  07  08  07  09  08  06  08  07               
3418  4  4     418  428  07  09  07  08  07  09  08  04  08  08               
3419  4  4     418  428  09  08  09  09  08  10  09  09  09  09               
3420  4  4     418  428  07  05  05  07  06  07  06  06  07  06               
3421     4     418  428  08  06  08  07  06  09  08  06  07  07               
3422  4  4     418  428  08  06  06  07  07  08  08  05  07  07               
3423  4  4     418  428  06  03  07  08  05  10  06  06  06  07               
...  .. .. ..  ...  ...  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..   
3479  4  4     417  427  06  04  04  04  05  08  06  07  03  05               
3480  4  4               10  10  10  09  08  10  10  10  10  10               
3481  4  4     417  427  10  10  10  10  10  10  10  10  10  10               
3482  4  4     417  427  08  07  08  07  07  08  08  08  08  07               
3483  4  4     417  427  09  08  08  08  07  09  09  09  09  09               
3484  4  4     417  427  10  10  09  08  08  10  10  09  10  09               
3485  4  4     417  427  08  08  07  07  06  10  10  10  10  08               
3486  4  4     417  427  10  10  10  08  10  10  10  10  10  10               
3487  4  4     417  427  09  06  05  05  09  10  09  09  09  07               
3488  4  4     417  427  09  07  10  05  02  09  10  09  09  08               
3489  4  4     417  427  10  08  09  08  04  09  09  08  08  08               
3490  4  4     417  427  09  08  08  07  05  09  09  09  08  08               
3491  4  4     417  427  07  08  07  04  06  09  09  09  07  07               
3492  4  4     417  427  03  03  02  02  03  06  05  04  01  02               
3493  4        417  427  05  03  03  04  05  07  06  07  07  03               
3494  4  4     417  427  08  06  07  04  07  10  07  09  06  06               
3495  4  4     417  427  07  08  07  05  06  08  07  07  08  07               
3496  4  4     417  427  08  08  07  07  06  10  08  07  08  08               
3497           417  427  06  01  04  05  04  10  05  10

Primero procederemos a rellenar el grupo de los registros correspondientes a postgrados, ya que los tenemos localizados, son justamente los registros debajo del curso sin grupo hasta el final del data frame

Estos son los registros a partir de la fila 3508

In [15]:
encu_1.iloc[3508:,2] = 'A'

In [18]:
encu_1.iloc[3508:,:]

0  1  2    3    4   5   6   7   8   9   10  11  12  13  14  15  16  17  \
3508  4  5  A  448  445  06  02  04  05  07  10  01  05  03  04               
3509  4  5  A  448  445  05  05  05  07  06  10  05  10  06  06               
3510  4  5  A  448  445  04  02  02  05  03  09  03  08  02  04               
3511     5  A  448  445  06  06  06  06  07  09  03  09  05  05               
3512  4  5  A  448  445  06  06  06  07  05  09  07  08  06  06               
3513  4  5  A  448  445  05  01  02  01  01  10  01  05  01  02               
3514  4  5  A  448  445  10  06  05  05  06  10  05  10  05  06               
3515  4  5  A  448  445  06  02  01  01  01  09  01  09  01  02               
3516  4  5  A  448  445  09  06  06  06  07  10  05  08  06  06               
3517  4  5  A  448  445  06  06  06  05  08  08      04  03  04               
3518  4  5  A  448  445  05  05  05  06  05  10  04  07  10  06               
3519  4  5  A  448  445  01  03  06  05  02  10  02  08  05  05               
3520  4  5  A  448  445  06  06  07  06  08  10  04  10  06  06               
3521  4  5  A  448  445  06  02  02  01  06  10  01  05  04  04               
3522  4  5  A  448  445  06  02  02  01  08  10  04  05  01  04               
3523  4     A  448  445  05  02  02  01  06  10  03  07  05  04               
3524     5  A  448  445  07  07  07  06  06  06  06  06  05  06               
3525     5  A  448  445  05  03  03  01  02  10  01  08  03  05               
3526  4  5  A  448  445  06  06  04  04  04  06  06  06  04  07               
3527  4  5  A  44   445  09  08  07  10  09  10  09  10  08  09               
3528  4  5  A  448  445  05  05  05  03  06  10  01  08  03  04               
3529  4  5  A  448  445  06  01  02  01  05  10  03  05  01  01               
3530  4  5  A  448  445  07  06  06  06  05  07  06  07  06  06               
3531  4  5  A  44   445  08  06  07  04  07  09  04  09  06  06               
3532  4  5  A  448  445  09  08  06  05  08  10  04  10  06  07               
3533  4  5  A  448  445  03  01  01  01  01  10  02  03  01  02               
3534  4  5  A  448  445  05  03  03  02  05  10  03  07  02  04               
3535  4  5  A  447  444  10  10  10  10  10  10  10  10  10  10               
3536  4  5  A  447  444  09  08  08  06  06  10  09  10  10  09               
3537  4  5  A  447  444  08  08  08  08  08  08  08  08  08  08               
...  .. .. ..  ...  ...  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..   
3699  1     A  253  254  07  08  07  08  07  07  08  08  07  07               
3700  1     A  253  254  08  07  07  07  07  07  07  07  07  07               
3701  1     A  253  254  02  03  04  04  04  04  05  05  06  05               
3702        A  253  254  09  08  07  09  08  09  09  09  08  08               
3703        A  253  254  09  09  08  09  08  09  09  09  09                   
3704        A  253  254  05  05  05  05  05  05  05  05  05  05               
3705  1     A  253  254  10  10  10  10  10  10  10  10  10  10               
3706        A  253  254  10  10  10  10  10  10  10  10  10  10               
3707  1     A  250  223  10  10  10  10  10  10  10  10  10  10               
3708  1     A  250  223  03  04  04  06  06  10  09  06  07  06               
3709  1     A  250  223  06  09  05  07  07  10  10          10  05           
3710        A  250  223  08  08  08  08  08  08  08  08  08  08               
3711  1     A  250  223  07  07  07  07  08  08  08  01  08  07               
3712        A  25   223  02  05  02  07  01  10  05  05  02  03               
3713  1     A  250  223  10  10  10  10  10  10  10  10  10  10               
3714  1     A  250  223  06  07  06  08  05  05  07  05  09  08               
3715        A  250  223  05  06  05  05  06  05  06  04  06  05               
3716  1     A  25   223  10  10  09  10  09  09  10  10  10  10               
3717  1     A  250  223  06  06  05  07  06  08  07  05

Ahora se eliminan los registros

In [19]:
encu_11 = encu_1.drop(encu_1.index[3395:3508])

In [20]:
encu_1.shape, encu_11.shape, encu_1.shape[0] - encu_11.shape[0]

((3729, 20), (3616, 20), 113)

Se han eliminado 113 filas

### Empieza el procesado!!

Se Cogen los items de respuesta (columnas de la 5 hasta el final) y con una expresión regular se sustituyen los espacios en blanco por NaN y luego se eliminan las columnas donde todos los valores son NaN's

In [21]:
encu_2 = encu_11.iloc[:,5:].replace(r'\s+', np.nan, regex=True).dropna(axis=1, how='all')
encu_2.head()

5   6   7   8   9   10  11   12  13  14   15   16
0  05  06  04  08  07  03  05   07  07  09  NaN  NaN
1  07  07  06  06  06  08  07  NaN  08  08  NaN  NaN
2  10  10  10  10  10  10  10   10  10  10  NaN  NaN
3  08  07  07  06  07  08  07  NaN  07  08  NaN  NaN
4  09  08  08  08  09  09  08   09  08  08  NaN  NaN

Obtenemos la primera versión reducida del proceso de limpieza. Nótese que estamos trabajando solo con los items de respuesta.

En este paso se deben realizar varias operaciones:

* Solo hay 10 items de respuesta (columnas 5 a 9), pero una vez eliminadas las columnas con todos los valores iguales a NaN aun
  quedan 2 columnas adicionales. Esto es debido a que la lectora ha desplazado algunas calificaciones hacia la derecha. Asi que 
  lo primero que haremos es corregir ese desplazamiento.
  
* Existen items sin calificar, a estos se le imputa el valor mediano del resto de items

Comenzamos con la primera operación, corregir el desplazamiento realizado por la lectora.

Se localizan los NaN's de la penúltima columna

In [22]:
i_nulos_p = encu_2.iloc[:,-2].isnull()

Lo anterior nos da un vector booleano de unos y ceros, los unos (TRUE) nos dan los elementos nulos.

En la próxima celda localizamos los valores no nulos buscando (con where) aquellas celdas iguales a cero

In [23]:
i_nonulos_p = np.where(i_nulos_p == 0)[0].tolist()
i_nonulos_p

[87,
 175,
 204,
 206,
 246,
 392,
 548,
 990,
 1269,
 1493,
 2000,
 2193,
 2480,
 2484,
 2563,
 2753,
 3035,
 3125,
 3231,
 3259,
 3596]

Repetimos lo anterior con la última columna

In [24]:
i_nulos_u = encu_2.iloc[:,-1].isnull()
i_nonulos_u = np.where(i_nulos_u == 0)[0].tolist()
i_nonulos_u

[175, 183, 302]

Se construye una lista que contenga los no nulos de la última y penúltima fila, usando la unión (``|``)

In [25]:
i_nonulos_temp = set(i_nonulos_p) | set(i_nonulos_u)
i_nonulos = sorted(list(i_nonulos_temp))
i_nonulos

[87,
 175,
 183,
 204,
 206,
 246,
 302,
 392,
 548,
 990,
 1269,
 1493,
 2000,
 2193,
 2480,
 2484,
 2563,
 2753,
 3035,
 3125,
 3231,
 3259,
 3596]

Se comprueba que la longitud es distinta a los nonulos de las dos últimas columnas

In [26]:
len(i_nonulos_p), len(i_nonulos_u), len(i_nonulos)

(21, 3, 23)

In [27]:
encu_2.iloc[i_nonulos,:]

5   6   7   8   9   10   11   12   13   14   15   16
87    10  08  06  07  09  05   10   04  NaN   07   07  NaN
175   08  04  06  07  07  06   05   04   06   05   04   04
183   05  03  05  08  08  10   10   09   09   06  NaN   06
204   07  06  05  07  08  10   08   06  NaN   06   07  NaN
206   04  06  04  06  08  09   07  NaN   06   04   07  NaN
246   07  03  05  08  04  08   07   08   05  NaN   07  NaN
302   10  09  10  07  10  10   10   07   05  NaN  NaN   08
392   07  06  08  06  07  01  NaN   07   06   07   08  NaN
548   09  06  09  08  08  07   09   05  NaN   09   07  NaN
990   08  10  10  08  06  04   08   08   10   09   09  NaN
1269  08  10  07  08  09  04   06  NaN   03   09   08  NaN
1493  06  06  06  07  07  06   07   08  NaN   06   07  NaN
2000  02  01  02  03  02  09   08   08   07  NaN   04  NaN
2193  07  03  05  05  04  10   10   08   04  NaN   08  NaN
2480  07  07  07  09  08  10   10   07   07   07   07  NaN
2484  04  02  03  02  04  10   04   07   05   04   04  NaN
2563  04  05  05  06  05  04  NaN   06   06   09   05  NaN
2753  04  03  03  03  06  08   06   04  NaN   04   02  NaN
3035  09  09  10  10  06  09   09   09   10  NaN   08  NaN
3125  08  02  03  06  05  10   09   07   02   04   04  NaN
3231  07  09  09  09  07  10   09   05   08   08   08  NaN
3259  10  09  09  09  09  10   10   10   10   09   09  NaN
3709  06  09  05  07  07  10   10  NaN  NaN   10   05  NaN

Se define una función que realiza lo siguiente:

- Verfifica si la fila tiene NaN's

- En caso afirmativo localiza los indices donde se encuentran los NaN's

- Si tiene un solo NaN (``if len (ii) < 2``) hace un loop que recorre la fila a partir de 
  dicho indice hasta la penultima posición, y sustituye cada *celda* por el valor de la siguiente
  
- En caso que tenga más de un NaN (``else``) hace lo mismo, pero sustituyendo la celda por el valor de
  de la celda + (número de nan's)

In [28]:
def llenar(x):
    if any(l == True for l in x[:-2].isnull()):
        ii = x[:-2].isnull()
        ii = np.where(ii == 1)[0].tolist()
        if len (ii) < 2:
            for indice in range(ii[0],(len(x)-1)):
                x.iloc[indice] = x.iloc[indice+1]
        else:
            for indice in range(ii[0],(len(x)-len(ii))):
                x.iloc[indice] = x.iloc[indice+len(ii)]
    return x

Se crea un data frame que contega solo los elementos arriba filtrados

In [29]:
prueba = encu_2.iloc[i_nonulos,:]
prueba

5   6   7   8   9   10   11   12   13   14   15   16
87    10  08  06  07  09  05   10   04  NaN   07   07  NaN
175   08  04  06  07  07  06   05   04   06   05   04   04
183   05  03  05  08  08  10   10   09   09   06  NaN   06
204   07  06  05  07  08  10   08   06  NaN   06   07  NaN
206   04  06  04  06  08  09   07  NaN   06   04   07  NaN
246   07  03  05  08  04  08   07   08   05  NaN   07  NaN
302   10  09  10  07  10  10   10   07   05  NaN  NaN   08
392   07  06  08  06  07  01  NaN   07   06   07   08  NaN
548   09  06  09  08  08  07   09   05  NaN   09   07  NaN
990   08  10  10  08  06  04   08   08   10   09   09  NaN
1269  08  10  07  08  09  04   06  NaN   03   09   08  NaN
1493  06  06  06  07  07  06   07   08  NaN   06   07  NaN
2000  02  01  02  03  02  09   08   08   07  NaN   04  NaN
2193  07  03  05  05  04  10   10   08   04  NaN   08  NaN
2480  07  07  07  09  08  10   10   07   07   07   07  NaN
2484  04  02  03  02  04  10   04   07   05   04   04  NaN
2563  04  05  05  06  05  04  NaN   06   06   09   05  NaN
2753  04  03  03  03  06  08   06   04  NaN   04   02  NaN
3035  09  09  10  10  06  09   09   09   10  NaN   08  NaN
3125  08  02  03  06  05  10   09   07   02   04   04  NaN
3231  07  09  09  09  07  10   09   05   08   08   08  NaN
3259  10  09  09  09  09  10   10   10   10   09   09  NaN
3709  06  09  05  07  07  10   10  NaN  NaN   10   05  NaN

Se aplica la función anterior a ``prueba``

In [30]:
prueba.apply(llenar, axis =1)

5   6   7   8   9   10  11  12  13   14   15   16
87    10  08  06  07  09  05  10  04  07   07  NaN  NaN
175   08  04  06  07  07  06  05  04  06   05   04   04
183   05  03  05  08  08  10  10  09  09   06  NaN   06
204   07  06  05  07  08  10  08  06  06   07  NaN  NaN
206   04  06  04  06  08  09  07  06  04   07  NaN  NaN
246   07  03  05  08  04  08  07  08  05   07  NaN  NaN
302   10  09  10  07  10  10  10  07  05  NaN   08   08
392   07  06  08  06  07  01  07  06  07   08  NaN  NaN
548   09  06  09  08  08  07  09  05  09   07  NaN  NaN
990   08  10  10  08  06  04  08  08  10   09   09  NaN
1269  08  10  07  08  09  04  06  03  09   08  NaN  NaN
1493  06  06  06  07  07  06  07  08  06   07  NaN  NaN
2000  02  01  02  03  02  09  08  08  07   04  NaN  NaN
2193  07  03  05  05  04  10  10  08  04   08  NaN  NaN
2480  07  07  07  09  08  10  10  07  07   07   07  NaN
2484  04  02  03  02  04  10  04  07  05   04   04  NaN
2563  04  05  05  06  05  04  06  06  09   05  NaN  NaN
2753  04  03  03  03  06  08  06  04  04   02  NaN  NaN
3035  09  09  10  10  06  09  09  09  10   08  NaN  NaN
3125  08  02  03  06  05  10  09  07  02   04   04  NaN
3231  07  09  09  09  07  10  09  05  08   08   08  NaN
3259  10  09  09  09  09  10  10  10  10   09   09  NaN
3709  06  09  05  07  07  10  10  10  05  NaN   05  NaN

In [31]:
prueba

5   6   7   8   9   10  11  12  13   14   15   16
87    10  08  06  07  09  05  10  04  07   07  NaN  NaN
175   08  04  06  07  07  06  05  04  06   05   04   04
183   05  03  05  08  08  10  10  09  09   06  NaN   06
204   07  06  05  07  08  10  08  06  06   07  NaN  NaN
206   04  06  04  06  08  09  07  06  04   07  NaN  NaN
246   07  03  05  08  04  08  07  08  05   07  NaN  NaN
302   10  09  10  07  10  10  10  07  05  NaN   08   08
392   07  06  08  06  07  01  07  06  07   08  NaN  NaN
548   09  06  09  08  08  07  09  05  09   07  NaN  NaN
990   08  10  10  08  06  04  08  08  10   09   09  NaN
1269  08  10  07  08  09  04  06  03  09   08  NaN  NaN
1493  06  06  06  07  07  06  07  08  06   07  NaN  NaN
2000  02  01  02  03  02  09  08  08  07   04  NaN  NaN
2193  07  03  05  05  04  10  10  08  04   08  NaN  NaN
2480  07  07  07  09  08  10  10  07  07   07   07  NaN
2484  04  02  03  02  04  10  04  07  05   04   04  NaN
2563  04  05  05  06  05  04  06  06  09   05  NaN  NaN
2753  04  03  03  03  06  08  06  04  04   02  NaN  NaN
3035  09  09  10  10  06  09  09  09  10   08  NaN  NaN
3125  08  02  03  06  05  10  09  07  02   04   04  NaN
3231  07  09  09  09  07  10  09  05  08   08   08  NaN
3259  10  09  09  09  09  10  10  10  10   09   09  NaN
3709  06  09  05  07  07  10  10  10  05  NaN   05  NaN

Sustituimos en el data frame original los valores modificados en el proceso anterior

In [32]:
encu_2.iloc[i_nonulos_p,:] = prueba
encu_2

5   6    7   8    9    10   11   12   13   14   15   16
0     05  06   04  08   07   03   05   07   07   09  NaN  NaN
1     07  07   06  06   06   08   07  NaN   08   08  NaN  NaN
2     10  10   10  10   10   10   10   10   10   10  NaN  NaN
3     08  07   07  06   07   08   07  NaN   07   08  NaN  NaN
4     09  08   08  08   09   09   08   09   08   08  NaN  NaN
5     07  08   08  06   09   09   05   06   07   09  NaN  NaN
6     06  07   10  10   08   09   10   10   10   09  NaN  NaN
7     08  09   07  09   04   07   09   10   10   09  NaN  NaN
8     07  02   07  08   08   06   09   08   02   07  NaN  NaN
9     09  08   08  07   08   07   09   08   09   09  NaN  NaN
10    08  05  NaN  03   06   09   09   09   04   08  NaN  NaN
11    05  09   05  10   09   10   10   10   09   06  NaN  NaN
12    10  10   10  10   10   10   10   10   10   10  NaN  NaN
13    07  06   07  08   07   06   ??   07   07   08  NaN  NaN
14    04  02   02  03   03   06   08   08   06   03  NaN  NaN
15    06  08   06  05   06  NaN   07   05   07   09  NaN  NaN
16    09  08   08  08   07   07   08   08   08   07  NaN  NaN
17    01  01   01  01   01   01   01   01   01   01  NaN  NaN
18    07  05   07  08   06   08   05   07   06   07  NaN  NaN
19    01  01   01  01   01   01   03   01   01   01  NaN  NaN
20    09  09   09  08   07   08   09   09   09   09  NaN  NaN
21    08  09   08  09   08   10   08   09   09   09  NaN  NaN
22    09  07   09  08   09   07   09   09   07   09  NaN  NaN
23    07  06   07  08   07   09   07   07   07   07  NaN  NaN
24    04  05   04  05   04   06   07   07   08   06  NaN  NaN
25    05  05   06  06   06   09   09   09   09   09  NaN  NaN
26    07  07   08  06   04   06   08   09   08   07  NaN  NaN
27    10  07   09  09   08   09   10   09   09   10  NaN  NaN
28    10  09   09  10   09   09   09   09   10   09  NaN  NaN
29    03  02   02  03   03   09   07   06   02   04  NaN  NaN
...   ..  ..  ...  ..  ...  ...  ...  ...  ...  ...  ...  ...
3699  07  08   07  08   07   07   08   08   07   07  NaN  NaN
3700  08  07   07  07   07   07   07   07   07   07  NaN  NaN
3701  02  03   04  04   04   04   05   05   06   05  NaN  NaN
3702  09  08   07  09   08   09   09   09   08   08  NaN  NaN
3703  09  09   08  09   08   09   09   09   09  NaN  NaN  NaN
3704  05  05   05  05   05   05   05   05   05   05  NaN  NaN
3705  10  10   10  10   10   10   10   10   10   10  NaN  NaN
3706  10  10   10  10   10   10   10   10   10   10  NaN  NaN
3707  10  10   10  10   10   10   10   10   10   10  NaN  NaN
3708  03  04   04  06   06   10   09   06   07   06  NaN  NaN
3709  06  09   05  07   07   10   10   10   05  NaN   05  NaN
3710  08  08   08  08   08   08   08   08   08   08  NaN  NaN
3711  07  07   07  07   08   08   08   01   08   07  NaN  NaN
3712  02  05   02  07   01   10   05   05   02   03  NaN  NaN
3713  10  10   10  10   10   10   10   10   10   10  NaN  NaN
3714  06  07   06  08   05   05   07   05   09   08  NaN  NaN
3715  05  06   05  05   06   05   06   04   06   05  NaN  NaN
3716  10  10   09  10   09   09   10   10   10   10  NaN  NaN
3717  06  06   05  07   06   08   07   05   05   05  NaN  NaN
3718  06  05   06  06   06   09   05   07   07   06  NaN  NaN
3719  02  ??   01  04  NaN   ??  NaN   07  NaN   06  NaN  NaN
3720  07  09   09  10   08   10   10   07   10   09  NaN  NaN
3721  07  07   07  07   07   07   07   07   07   07  NaN  NaN
3722  05  05   06  07   06   08   07   05   05   06  NaN  NaN
3723  04  04   04  04   04   04   05   03   04   05  NaN  NaN
3724  08  08   07  08   08   08   09   07   08   08  NaN  NaN
3725  05  05   04  05   04   05   05   05   05   05  NaN  NaN
3726  04  04   04  05   04   07   05   03   03   04  NaN  NaN
3727  05  04   04  05   06   06   07   09   08   07  NaN  NaN
3728  07  07   06  09   04   09   08   09   07  NaN  NaN  NaN

[3616 rows x 12 columns]

Eliminamos las dos últimas columnas

In [33]:
encu_2.drop(encu_2.columns[-2:],inplace=True,axis=1)

In [34]:
encu_2

5   6    7   8    9    10   11   12   13   14
0     05  06   04  08   07   03   05   07   07   09
1     07  07   06  06   06   08   07  NaN   08   08
2     10  10   10  10   10   10   10   10   10   10
3     08  07   07  06   07   08   07  NaN   07   08
4     09  08   08  08   09   09   08   09   08   08
5     07  08   08  06   09   09   05   06   07   09
6     06  07   10  10   08   09   10   10   10   09
7     08  09   07  09   04   07   09   10   10   09
8     07  02   07  08   08   06   09   08   02   07
9     09  08   08  07   08   07   09   08   09   09
10    08  05  NaN  03   06   09   09   09   04   08
11    05  09   05  10   09   10   10   10   09   06
12    10  10   10  10   10   10   10   10   10   10
13    07  06   07  08   07   06   ??   07   07   08
14    04  02   02  03   03   06   08   08   06   03
15    06  08   06  05   06  NaN   07   05   07   09
16    09  08   08  08   07   07   08   08   08   07
17    01  01   01  01   01   01   01   01   01   01
18    07  05   07  08   06   08   05   07   06   07
19    01  01   01  01   01   01   03   01   01   01
20    09  09   09  08   07   08   09   09   09   09
21    08  09   08  09   08   10   08   09   09   09
22    09  07   09  08   09   07   09   09   07   09
23    07  06   07  08   07   09   07   07   07   07
24    04  05   04  05   04   06   07   07   08   06
25    05  05   06  06   06   09   09   09   09   09
26    07  07   08  06   04   06   08   09   08   07
27    10  07   09  09   08   09   10   09   09   10
28    10  09   09  10   09   09   09   09   10   09
29    03  02   02  03   03   09   07   06   02   04
...   ..  ..  ...  ..  ...  ...  ...  ...  ...  ...
3699  07  08   07  08   07   07   08   08   07   07
3700  08  07   07  07   07   07   07   07   07   07
3701  02  03   04  04   04   04   05   05   06   05
3702  09  08   07  09   08   09   09   09   08   08
3703  09  09   08  09   08   09   09   09   09  NaN
3704  05  05   05  05   05   05   05   05   05   05
3705  10  10   10  10   10   10   10   10   10   10
3706  10  10   10  10   10   10   10   10   10   10
3707  10  10   10  10   10   10   10   10   10   10
3708  03  04   04  06   06   10   09   06   07   06
3709  06  09   05  07   07   10   10   10   05  NaN
3710  08  08   08  08   08   08   08   08   08   08
3711  07  07   07  07   08   08   08   01   08   07
3712  02  05   02  07   01   10   05   05   02   03
3713  10  10   10  10   10   10   10   10   10   10
3714  06  07   06  08   05   05   07   05   09   08
3715  05  06   05  05   06   05   06   04   06   05
3716  10  10   09  10   09   09   10   10   10   10
3717  06  06   05  07   06   08   07   05   05   05
3718  06  05   06  06   06   09   05   07   07   06
3719  02  ??   01  04  NaN   ??  NaN   07  NaN   06
3720  07  09   09  10   08   10   10   07   10   09
3721  07  07   07  07   07   07   07   07   07   07
3722  05  05   06  07   06   08   07   05   05   06
3723  04  04   04  04   04   04   05   03   04   05
3724  08  08   07  08   08   08   09   07   08   08
3725  05  05   04  05   04   05   05   05   05   05
3726  04  04   04  05   04   07   05   03   03   04
3727  05  04   04  05   06   06   07   09   08   07
3728  07  07   06  09   04   09   08   09   07  NaN

[3616 rows x 10 columns]

Ahora procedemos a la siguiente operación, imputar los items vacios por el valor mediano del resto de las valoraciones (fila)

Primero sustituimos los '??' por NAN

In [35]:
encu_3 = encu_2.replace(['??'], np.nan)

In [36]:
encu_3

5    6    7   8    9    10   11   12   13   14
0     05   06   04  08   07   03   05   07   07   09
1     07   07   06  06   06   08   07  NaN   08   08
2     10   10   10  10   10   10   10   10   10   10
3     08   07   07  06   07   08   07  NaN   07   08
4     09   08   08  08   09   09   08   09   08   08
5     07   08   08  06   09   09   05   06   07   09
6     06   07   10  10   08   09   10   10   10   09
7     08   09   07  09   04   07   09   10   10   09
8     07   02   07  08   08   06   09   08   02   07
9     09   08   08  07   08   07   09   08   09   09
10    08   05  NaN  03   06   09   09   09   04   08
11    05   09   05  10   09   10   10   10   09   06
12    10   10   10  10   10   10   10   10   10   10
13    07   06   07  08   07   06  NaN   07   07   08
14    04   02   02  03   03   06   08   08   06   03
15    06   08   06  05   06  NaN   07   05   07   09
16    09   08   08  08   07   07   08   08   08   07
17    01   01   01  01   01   01   01   01   01   01
18    07   05   07  08   06   08   05   07   06   07
19    01   01   01  01   01   01   03   01   01   01
20    09   09   09  08   07   08   09   09   09   09
21    08   09   08  09   08   10   08   09   09   09
22    09   07   09  08   09   07   09   09   07   09
23    07   06   07  08   07   09   07   07   07   07
24    04   05   04  05   04   06   07   07   08   06
25    05   05   06  06   06   09   09   09   09   09
26    07   07   08  06   04   06   08   09   08   07
27    10   07   09  09   08   09   10   09   09   10
28    10   09   09  10   09   09   09   09   10   09
29    03   02   02  03   03   09   07   06   02   04
...   ..  ...  ...  ..  ...  ...  ...  ...  ...  ...
3699  07   08   07  08   07   07   08   08   07   07
3700  08   07   07  07   07   07   07   07   07   07
3701  02   03   04  04   04   04   05   05   06   05
3702  09   08   07  09   08   09   09   09   08   08
3703  09   09   08  09   08   09   09   09   09  NaN
3704  05   05   05  05   05   05   05   05   05   05
3705  10   10   10  10   10   10   10   10   10   10
3706  10   10   10  10   10   10   10   10   10   10
3707  10   10   10  10   10   10   10   10   10   10
3708  03   04   04  06   06   10   09   06   07   06
3709  06   09   05  07   07   10   10   10   05  NaN
3710  08   08   08  08   08   08   08   08   08   08
3711  07   07   07  07   08   08   08   01   08   07
3712  02   05   02  07   01   10   05   05   02   03
3713  10   10   10  10   10   10   10   10   10   10
3714  06   07   06  08   05   05   07   05   09   08
3715  05   06   05  05   06   05   06   04   06   05
3716  10   10   09  10   09   09   10   10   10   10
3717  06   06   05  07   06   08   07   05   05   05
3718  06   05   06  06   06   09   05   07   07   06
3719  02  NaN   01  04  NaN  NaN  NaN   07  NaN   06
3720  07   09   09  10   08   10   10   07   10   09
3721  07   07   07  07   07   07   07   07   07   07
3722  05   05   06  07   06   08   07   05   05   06
3723  04   04   04  04   04   04   05   03   04   05
3724  08   08   07  08   08   08   09   07   08   08
3725  05   05   04  05   04   05   05   05   05   05
3726  04   04   04  05   04   07   05   03   03   04
3727  05   04   04  05   06   06   07   09   08   07
3728  07   07   06  09   04   09   08   09   07  NaN

[3616 rows x 10 columns]

Se define una función que calcula la mediana de cada fila y porteriormente sustituye los nan's por dicho valor

In [37]:
def reemplazo_mediana(x):
    mediana = x.median()
    return x.fillna(mediana)

Se aplica la función anterior a las filas del data frame

In [38]:
encu_3.apply(reemplazo_mediana, axis = 1)

5   6   7   8   9   10  11  12  13  14
0     05  06  04  08  07  03  05  07  07  09
1     07  07  06  06  06  08  07   7  08  08
2     10  10  10  10  10  10  10  10  10  10
3     08  07  07  06  07  08  07   7  07  08
4     09  08  08  08  09  09  08  09  08  08
5     07  08  08  06  09  09  05  06  07  09
6     06  07  10  10  08  09  10  10  10  09
7     08  09  07  09  04  07  09  10  10  09
8     07  02  07  08  08  06  09  08  02  07
9     09  08  08  07  08  07  09  08  09  09
10    08  05   8  03  06  09  09  09  04  08
11    05  09  05  10  09  10  10  10  09  06
12    10  10  10  10  10  10  10  10  10  10
13    07  06  07  08  07  06   7  07  07  08
14    04  02  02  03  03  06  08  08  06  03
15    06  08  06  05  06   6  07  05  07  09
16    09  08  08  08  07  07  08  08  08  07
17    01  01  01  01  01  01  01  01  01  01
18    07  05  07  08  06  08  05  07  06  07
19    01  01  01  01  01  01  03  01  01  01
20    09  09  09  08  07  08  09  09  09  09
21    08  09  08  09  08  10  08  09  09  09
22    09  07  09  08  09  07  09  09  07  09
23    07  06  07  08  07  09  07  07  07  07
24    04  05  04  05  04  06  07  07  08  06
25    05  05  06  06  06  09  09  09  09  09
26    07  07  08  06  04  06  08  09  08  07
27    10  07  09  09  08  09  10  09  09  10
28    10  09  09  10  09  09  09  09  10  09
29    03  02  02  03  03  09  07  06  02  04
...   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..
3699  07  08  07  08  07  07  08  08  07  07
3700  08  07  07  07  07  07  07  07  07  07
3701  02  03  04  04  04  04  05  05  06  05
3702  09  08  07  09  08  09  09  09  08  08
3703  09  09  08  09  08  09  09  09  09   9
3704  05  05  05  05  05  05  05  05  05  05
3705  10  10  10  10  10  10  10  10  10  10
3706  10  10  10  10  10  10  10  10  10  10
3707  10  10  10  10  10  10  10  10  10  10
3708  03  04  04  06  06  10  09  06  07  06
3709  06  09  05  07  07  10  10  10  05   7
3710  08  08  08  08  08  08  08  08  08  08
3711  07  07  07  07  08  08  08  01  08  07
3712  02  05  02  07  01  10  05  05  02  03
3713  10  10  10  10  10  10  10  10  10  10
3714  06  07  06  08  05  05  07  05  09  08
3715  05  06  05  05  06  05  06  04  06  05
3716  10  10  09  10  09  09  10  10  10  10
3717  06  06  05  07  06  08  07  05  05  05
3718  06  05  06  06  06  09  05  07  07  06
3719  02   4  01  04   4   4   4  07   4  06
3720  07  09  09  10  08  10  10  07  10  09
3721  07  07  07  07  07  07  07  07  07  07
3722  05  05  06  07  06  08  07  05  05  06
3723  04  04  04  04  04  04  05  03  04  05
3724  08  08  07  08  08  08  09  07  08  08
3725  05  05  04  05  04  05  05  05  05  05
3726  04  04  04  05  04  07  05  03  03  04
3727  05  04  04  05  06  06  07  09  08  07
3728  07  07  06  09  04  09  08  09  07   7

[3616 rows x 10 columns]

Ya tenemos los valores correspondientes a los items de valoración *limpios*